Refactor nccl_ofi_idpool to C++

AvivBenchorin opened the pull request:
Refactors the nccl_ofi_idpool implementation to C++, creating a nccl_ofi_idpool_t class with std::vector and std::mutex member variables and thread-safe wrappers to interact with the std:vector. Updates all usage of the previous nccl_ofi_idpool C implementation to use the new class.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

bwbarrett commented on include/nccl_ofi_idpool.h:
```diff
@@ -5,90 +5,77 @@
+class nccl_ofi_idpool_t {
+public:
+	/* Disable default constructor to require passing in a size */
+	nccl_ofi_idpool_t() = delete;
```
Since you provide the constructor below, you shouldn't need to explicitly delete the default constructor.  I think.

On the other hand, we probably do want to delete the implicit copy constructor and just not have one.  I'm not sure I can come up with reasonable copy semantics for an id pool without way overcomplicating the id pool.

AvivBenchorin replied:
I tested it and you are correct, I wouldn't need to explicitly delete the default constructor. And makes sense for the implicit copy constructor, I'll delete it along with the copy assignment operator:
```
nccl_ofi_idpool_t(const nccl_ofi_idpool_t&) = delete;
nccl_ofi_idpool_t& operator=(const nccl_ofi_idpool_t&) = delete;
```

AvivBenchorin replied:
Changed in force push.

bwbarrett commented on include/nccl_ofi_idpool.h:
```diff
@@ -5,90 +5,77 @@
+	 * @return	the extracted ID (zero-based) on success,
+	 *		negative value on error
+	 */
+	int allocate_id();
```
I know these were ints, but we should use ssize_t here.

bwbarrett commented on include/nccl_ofi_idpool.h:
```diff
@@ -5,90 +5,77 @@
+	 * @brief	Allocate an ID
+	 *
+	 * Extract an available ID from the ID pool, mark the ID as
+	 * unavailable in the pool, and return extracted ID. No-op in case
```
It's not entirely a no-op, in that an error is returned :)

bwbarrett commented on include/nccl_ofi_idpool.h:
```diff
@@ -5,90 +5,77 @@
+	 * @return	0 on success
+	 *		non-zero on error
+	 */
+	int free_id(const size_t& id);
```
same POD argument as the constructor.

bwbarrett commented on include/nccl_ofi_idpool.h:
```diff
@@ -5,90 +5,77 @@
+
+private:
+	/* Size of the id pool (number of IDs) */
+	size_t size_;
```
We should be consistent in member names.  Either they all have trailing underscore or none do.  I'm a fan of the none option.

AvivBenchorin replied:
Fixed in force push, I removed all trailing underscores.

bwbarrett commented on src/nccl_ofi_idpool.cpp:
```diff
@@ -4,116 +4,53 @@
+	size_(size)
+{
+	if (size_ == 0) {
+		idpool_ = std::vector<uint64_t>();
```
I'd just default construct the vector in either case and resize it if size is not 0.

AvivBenchorin replied:
Fixed in force push.

bwbarrett commented on src/nccl_ofi_idpool.cpp:
```diff
@@ -4,116 +4,53 @@
-		free(idpool->ids);
-		idpool->ids = NULL;
-		return ret;
+	if (size_ % 8 != 0) {
```
throw a comment in here about what this is doing.

AvivBenchorin replied:
Added in force push, although my comment may need a revision for clarity.

bwbarrett commented on src/nccl_ofi_idpool.cpp:
```diff
@@ -4,116 +4,53 @@
+
+	/* Iterate over each of the idpool_ vector elements */
+	for (size_t i = 0; i < idpool_.size(); i++) {
+		entry_index = __builtin_ffsll(idpool_.at(i));
```
it's a vector.  I'd just write this and line 53 below as `idpool[i]`.

AvivBenchorin replied:
Fixed in force push. I had originally used `.at` since it has built-in bound-checking, but it would make sense with our current exception handling that we would rather then the bounds ourselves instead of throwing an exception.

AvivBenchorin commented on include/nccl_ofi_idpool.h:
```diff
@@ -5,90 +5,77 @@
 
-#include <pthread.h>
-#include <stdint.h>
+#include <algorithm>
```
Fixed in latest force push!

AvivBenchorin commented on src/nccl_ofi_idpool.cpp:
```diff
@@ -4,116 +4,70 @@
+#include <mutex>
+#include <vector>
+
+#include <unistd.h>
```
In the testing for this PR, I had ran into an issue where `ssize_t` was not defined on Ubuntu 22.04 without including `unistd.h` despite it being defined on AL2/AL2023. Given this, would it be appropriate for me to still include `errno.h` for the error codes and  `stdint.h` for `uint64_t` in case an OS would not have those definitions by default?

AvivBenchorin replied:
Looking at the `ssize_t` build failure again [here](https://github.com/aws/aws-ofi-nccl/actions/runs/14180354608/job/39724851231), the difference for whether `ssize_t` was defined without including `unistd.h` was not between Ub22 vs AL2/AL2023, and instead was between which compiler and compiler version was used on Ub22 (`gcc-latest` failed the build, while `gcc-legacy`/`clang-legacy`/`clang-latest` succeeded).

bwbarrett commented on include/nccl_ofi_idpool.h:
```diff
@@ -5,90 +5,78 @@
+
+	/* Return the idpool element value of a valid vector index, return 
+	   negative error code on out-of-bound index */
+	uint64_t get_element(size_t index);
```
I'm not a huge fan of making public interface functions just for testing.  I think instead I'd make the variables protected instead of private, and write a class in your test function that inherits from idpool and provides accessors to idpool.  That way, you don't have this weird interface that confuses everyone.

AvivBenchorin commented on include/nccl_ofi_idpool.h:
```diff
@@ -5,90 +5,78 @@
+	 * @return	0 on success
+	 *		non-zero on error
+	 */
+	int free_id(size_t id);
```
That's a good point, I agree with just throwing an exception. Currently in `nccl_ofi_idpool_t::free_id` we error if:
1. the idpool has a size of 0
2. the provided `id` to free is out of range
3. the provided `id` to free is not already in use
In all three situations something has gone wrong enough that would warrant aborting the program. In (1) we're trying to interact with an idpool that was explicitly created with a size of 0 and isn't intended to be used, and (2) and (3) could indicate either that `allocate_id` is returning invalid ids or not properly updating the idpool, or that the idpool is not thread-safe and the same id is getting free'd multiple times.

Something to note, in almost all of the `free_id` calls in the plugin, the error handling of the `ret` value is not short-circuiting to clean-up or returning the error code (e.g. in `nccl_ofi_rdma.cpp`, `dereg_mr`:
```
if (key_pool->get_size() != 0) {
	ret = key_pool->free_id(mr_handle->mr_key);
	if (OFI_UNLIKELY(ret != 0)) {
		NCCL_OFI_WARN("Error freeing MR key %ld, leaking key", mr_handle->mr_key);
	}
}
```
)
So with this change to throwing exceptions, we will now abort the program on these failures rather than just printing a warning and continuing.

AvivBenchorin replied:
In this same vein, I could update the `nccl_ofi_idpool_t::allocate_id` function to also throw exceptions rather than return an error code, it fails when:
1. The idpool has a size of 0
2. No IDs were available in the pool
Both of which would also be catastrophic failures (especially given we dont haven't resizing the pool to be larger in the case of (2)).

AvivBenchorin replied:
In the current code the only call to `free_id` that would actually use the return code is in [`send_comm_destroy`](https://github.com/aws/aws-ofi-nccl/blob/9761e70e1281559a84ef2c21fd82ffcf0193d861/src/nccl_ofi_rdma.cpp#L3968):
```
/* Release communicator ID */
ret = nccl_ofi_idpool_free_id(device->comm_idpool, s_comm->local_comm_id);
if (OFI_UNLIKELY(ret != 0)) {
	NCCL_OFI_WARN("Error freeing communicator ID %" PRIu32, s_comm->local_comm_id);
	return ret;
}
```
Which I think could be appropriate to convert from returning to just leaking if we wanted to avoid the exceptions. For my understanding, would there be a particular reason for it to be better to leak with `free_id` rather than throwing exceptions? I still feel that something going wrong with `free_id` would justify aborting the program.

For `allocate_id`, we could return `COMM_ID_INVALID` if the size is full for the communicator idpools. I couldn't find `FI_MR_NOTAVAIL`, but would [`FI_KEY_NOTAVAIL`](https://github.com/ofiwg/libfabric/blob/e3d9f1a36db1dc5fb7018b0cf0586ac4ff93fcd2/include/rdma/fabric.h#L220C9-L220C24) be appropriate for the MR use case?

bwbarrett replied:
`FI_KEY_NOTAVAIL` is what I meant, yeah.

I think throwing is fine.  I have no strong opinions either way on these.

AvivBenchorin commented:
Notes for latest PR force push:
1. I added a new commit [Fix setting return code in idpool error handling](https://github.com/aws/aws-ofi-nccl/pull/830/commits/7f7e00e59a5a2b5952e57c0deb39721635258ff9) before the idpool C++ migration commit which fixes, what I see as a bug in the error handling of some calls to `nccl_ofi_idpool_allocate_id`. I tried to have the bugfix first so that in case the C++ idpool commit needs to get reverted, the fix could stay. Please call out if this is not appropriate and if the bugfix should be a separate PR, or the two commits should be squashed so that the fix is a part of the C++ migration.
2. For the exception throws I added, I kept the exception message empty and instead kept the error message in the `NCCL_OFI_WARN` call. If it would be better to have the exception message also contain the same log as the `NCCL_OFI_WARN` call or should replace the logging call, I can make that change.
3. I have `nccl_ofi_idpool_t::allocate_id` returning `FI_KEY_NOTAVAIL` when the idpool is full, which is then used in error handling. However I am also reusing the `COMM_ID_INVALID` value for the error handling with `s_comm`/`r_comm`/`l_comm` comm_ids since the `uint64_t` sized `FI_KEY_NOTAVAIL`  overflows when being assigned to the `uint32_t` comm_id. If we want to just use `FI_KEY_NOTAVAIL` I can add static casts to `uint32_t` for `FI_KEY_NOTAVAIL` to avoid the overflow.

bwbarrett commented on src/nccl_ofi_idpool.cpp:
```diff
@@ -122,100 +78,45 @@ int nccl_ofi_idpool_allocate_id(nccl_ofi_idpool_t *idpool)
-		return -EINVAL;
+	if (OFI_UNLIKELY(id >= size)) {
+		NCCL_OFI_WARN("ID value %lu out of range (max: %lu)", id, size);
+		throw std::runtime_error("");
```
we can be a bit more helpful than an empty string :).

AvivBenchorin replied:
Added more descriptive exception messages in latest force push.

AvivBenchorin commented:
Rebased on top of master due to minor conflicts with #833. Added proper error messages to thrown exceptions.
